services/wallet/activity/service.go

```diff
@@ -0,0 +1,123 @@
+package activity
+
+import (
+	"context"
+	"database/sql"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/event"
+	"github.com/ethereum/go-ethereum/log"
+
+	w_common "github.com/status-im/status-go/services/wallet/common"
+	"github.com/status-im/status-go/services/wallet/walletevent"
+)
+
+const (
+	// FilterResponse json is sent as a message in the EventActivityFilteringDone event
+	EventActivityFilteringDone walletevent.EventType = "wallet-activity-filtering-done"
+)
+
+type Service struct {
+	db        *sql.DB
+	eventFeed *event.Feed
+
+	context  context.Context
+	cancelFn context.CancelFunc
+	wg       sync.WaitGroup
+	mu       sync.Mutex
+}
+
+func NewService(db *sql.DB, eventFeed *event.Feed) *Service {
+	return &Service{
+		db:        db,
+		eventFeed: eventFeed,
+	}
+}
+
+type ErrorCode = int
+
+const (
+	ErrorCodeSuccess ErrorCode = iota + 1
+	ErrorCodeFilterCanceled
+	ErrorCodeFilterFailed
+	ErrorCodeServiceFailed
+)
+
+type FilterResponse struct {
+	Activities       []Entry   `json:"activities"`
+	ThereMightBeMore bool      `json:"thereMightBeMore"`
+	ErrorCode        ErrorCode `json:"errorCode"`
+}
+
+// FilterActivityAsync allows only one filter task to run at a time
+// and it cancels the current one if a new one is started
+// All calls will trigger an EventActivityFilteringDone event with the result of the filtering
+func (s *Service) FilterActivityAsync(ctx context.Context, addresses []common.Address, chainIDs []w_common.ChainID, filter Filter, offset int, limit int) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	// If a previous task is running, cancel it and wait to finish
+	if s.cancelFn != nil {
+		s.cancelFn()
+		s.wg.Wait()
+	}
+
+	if ctx.Err() == nil {
+		s.context, s.cancelFn = context.WithCancel(context.Background())
```

Comment:
It would be nice to address this corner case. I would like to wait and see if this is a real issue in practice. Moving forward, we will index all the filterable columns and extract all required metadata in the same async call. This should speed up the process and cover the most extreme corner case.

Regarding UX, the user will see loading animations and switching from the activity tab or on filter reset the current long call will be aborted.

I would suggest waiting for field testing and seeing if this is a genuine concern before addressing it.

Reply:
I was including stability concerns in the performance case, also. However, if the DB contention is too high or we have unexpected behaviour, we must address it instead of reporting the error to the user.

> ... about other issues that we might have with DB.
Without a specific use case, it looks like premature optimization. What do you think?

Reply:
I agree with you that it will be ideal to handle as many corner cases as possible. The question where do you draw the line? In this case, waiting for a filter query for more than expected is a good enough indication that something is at fault. For further escalation, we will see in the API logs that a filter was called, which never received a response.

> Regarding premature optimization. Here is the case - currently we have not even switched to a separate DB, so we can be all stuck if messenger team or one of our requests suddenly takes the write lock for long. Again, if users catch it, and see the error they can report it.

This is a specific generic DB performance problem that is in work already. AFAIK we have tried to enable concurrent calls optimization that didn't work and next, there is a plan to address it as you suggested.

> Well, as I see it, we should not rely on being able to catch all issues but have them all logged and propagated, when it matters for users like why they are waiting forever to filter the activities.

What do you think will be a good timeout for this DB call? Which is the extreme expected data corner case and how long will the user expect to wait for a result?

Reply:
Hey @IvanBelyakoff, created this [issue](https://github.com/status-im/status-desktop/issues/11036) to track and address the concerns raised by this conversation. Let's continue our discussion in there.